File: reader.js
// @wolfram77


// required modules
var Gpio = require('onoff').Gpio;


// config
var ctimeout = 100;


// connections
var pdata0 = new Gpio(14, 'in', 'falling');
var pdata1 = new Gpio(15, 'in', 'falling');
// var pgreen = 18
// var pred = 23
// var pbeep = 24
// var phold = 25
// var pcard = 8


// initialize
var card = 0
var cbits = 0



// handle card data0 low
pdata0.watch(function(err, val) {
  card <<= 1
  cbits += 1
});


// handle card data1 low
pdata1.watch(function(err, val) {
  card = (card << 1) | 1
  cbits += 1
});


// card read code
console.log('pi-snax-reader');
setInterval(function() {
  if(cbits > 0) setTimeout(function() {
    console.log('['+cbits+'] - '+card);
    card = cbits = 0;
  }, ctimeout);
}, ctimeout);


// cleanup
process.on('SIGINT', function() {
  pdata0.unexport();
  pdata1.unexport();
  process.exit();
});

File: modules/reader.py
#!/usr/bin/env python
# @wolfram77


# required modules
import time
import RPi.GPIO as GPIO
from requests_futures.sessions import FuturesSession


# config
ctimeout = 0.1


# connections
pdata0 = 14
pdata1 = 15
pgreen = 18
pred = 23
pbeep = 24
phold = 25
pcard = 8


# pin modes
GPIO.setmode(GPIO.BCM)
GPIO.setup(pdata0, GPIO.IN)
GPIO.setup(pdata1, GPIO.IN)


# initialize
session = FuturesSession()
card = 1
cbits = 0



# handle card data0 low
def cdata0_low(channel):
  global card, cbits
  card <<= 1
  cbits += 1


# handle card data1 low
def cdata1_low(channel):
  global card, cbits
  card = (card << 1) | 1
  cbits += 1


# set card data handlers
GPIO.add_event_detect(pdata0, GPIO.FALLING, callback=cdata0_low)
GPIO.add_event_detect(pdata1, GPIO.FALLING, callback=cdata1_low)


# card read code
print "[reader] ready!"
try:
  while True:
    time.sleep(ctimeout)
    if cbits > 0:
      time.sleep(ctimeout)
      print "[%d bit] - %d" % (cbits, card)
      payload = {"card": card, "cbits": cbits}
      session.post("http://localhost/api/reader/card", payload)
      card = 1
      cbits = 0
except KeyboardInterrupt:
  GPIO.cleanup()
